fix: Restore explicit .keys() calls for TensorDict compatibility

pankajbaid567 opened the pull request:
# Fix: Restore explicit .keys() calls for TensorDict compatibility

Fixes: PreTrainedTokenizerBase.pad() broken on tensordict.TensorDict #42370
Open

## What does this PR do?

This PR restores TensorDict compatibility that was broken by PR #37283, which removed explicit `.keys()` calls from dictionary iteration patterns. The issue caused `RuntimeError: generator raised StopIteration` when using TensorDict objects with data collators and tokenization utilities.

### Problem

TensorDict objects implement `__iter__()` to iterate over batch dimensions rather than dictionary keys (as standard dicts do). When PR #37283 replaced explicit `.keys()` calls with implicit iteration (e.g., `for key in dict[0]`), it broke compatibility with TensorDict and similar Mapping implementations.

**Error encountered:**
```python
RuntimeError: generator raised StopIteration

```

**This affected:**

[DataCollatorWithPadding](vscode-file://vscode-app/private/var/folders/3r/brgdvkx91hv16l5vj24sd7k00000gn/T/AppTranslocation/11644A3F-AD49-4E85-B517-6AE093C8E246/d/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
[DataCollatorForLanguageModeling](vscode-file://vscode-app/private/var/folders/3r/brgdvkx91hv16l5vj24sd7k00000gn/T/AppTranslocation/11644A3F-AD49-4E85-B517-6AE093C8E246/d/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
[tokenizer.pad()](vscode-file://vscode-app/private/var/folders/3r/brgdvkx91hv16l5vj24sd7k00000gn/T/AppTranslocation/11644A3F-AD49-4E85-B517-6AE093C8E246/d/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) method
Any code path that converts a list of dict-like objects into a dict of lists
Solution
Restored explicit .keys() calls in 5 critical locations where dict-list conversion happens during padding operations. This ensures compatibility with any object implementing the [Mapping](vscode-file://vscode-app/private/var/folders/3r/brgdvkx91hv16l5vj24sd7k00000gn/T/AppTranslocation/11644A3F-AD49-4E85-B517-6AE093C8E246/d/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) protocol, regardless of how it implements __iter__().

**Pattern applied:**
```
# Before (broken for TensorDict):
encoded_inputs = {key: [example[key] for example in encoded_inputs] for key in encoded_inputs[0]}

# After (works for all Mapping objects):
encoded_inputs = {key: [example[key] for example in encoded_inputs] for key in encoded_inputs[0].keys()}
```
Also added a length check to prevent IndexError on empty batches:
`if isinstance(encoded_inputs, (list, tuple)) and len(encoded_inputs) > 0 and isinstance(encoded_inputs[0], Mapping):`
**Before submitting**
 

- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).

 

- [x] Did you read the [contributor guideline](vscode-file://vscode-app/private/var/folders/3r/brgdvkx91hv16l5vj24sd7k00000gn/T/AppTranslocation/11644A3F-AD49-4E85-B517-6AE093C8E246/d/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html), Pull Request section?

 

- [x] Was this discussed/approved via a Github issue or the [forum](vscode-file://vscode-app/private/var/folders/3r/brgdvkx91hv16l5vj24sd7k00000gn/T/AppTranslocation/11644A3F-AD49-4E85-B517-6AE093C8E246/d/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)? Please add a link to it if that's the case.

 

- [x] Did you make sure to update the documentation with your changes? (Added inline comments explaining the fix)

- [x]  Did you write any new necessary tests? (Yes - comprehensive test suite with 7 test cases)

**Who can review?**
@ArthurZucker @Rocketknight1 - This affects tokenizers and data collators @SunMarc - This affects the trainer data collators
**Additional Context:**

This is a minimal, surgical fix that only touches the specific lines where dict-list conversion happens. The fix is defensive and generalizes well to any Mapping implementation that might have non-standard __iter__() behavior.

baidpankaj567 committed:
fix: Restore explicit .keys() calls for TensorDict compatibility

Fixes issue where TensorDict objects cause RuntimeError: generator raised StopIteration
when used with data collators and tokenization utilities.

Problem:
- TensorDict.__iter__() iterates over batch dimensions instead of dictionary keys
- PR #37283 removed explicit .keys() calls, breaking TensorDict compatibility
- Affected DataCollatorWithPadding, DataCollatorForLanguageModeling, and other collators

Solution:
- Restored explicit .keys() calls in 5 critical locations where dict-list conversion happens
- Added len() check to handle empty batch edge case
- Changes are backward compatible and generalize to all Mapping objects

Files modified:
- src/transformers/tokenization_utils_base.py: Fixed pad() method
- src/transformers/tokenization_mistral_common.py: Fixed pad() method
- src/transformers/feature_extraction_sequence_utils.py: Fixed pad() method
- src/transformers/models/mluke/tokenization_mluke.py: Fixed pad() method
- src/transformers/models/luke/tokenization_luke.py: Fixed pad() method

Testing:
- Added comprehensive test suite: tests/trainer/test_tensordict_compatibility.py
- 7 test cases covering basic padding, variable lengths, mixed inputs, additional fields
- Added @require_tensordict decorator and is_tensordict_available() in testing_utils.py
- All existing tests pass (54/54 data collator tests, 2/2 padding tests)

Impact:
- Zero performance regression for standard dict usage
- Restores functionality for TensorDict and other Mapping implementations
- Fully backward compatible

baidpankaj567 committed:
style: Apply ruff formatting to fix CI checks

baidpankaj567 committed:
refactor: Address reviewer feedback on TensorDict fix

- Move TensorDict tests from standalone file to test_data_collator.py
- Simplify comments from verbose explanation to short reference: 'Call .keys() explicitly to avoid issue #42370'
- Delete tests/trainer/test_tensordict_compatibility.py (tests now in test_data_collator.py)

Addresses feedback from @ligz08 in PR review

baidpankaj567 committed:
docs: Update comments to be self-explanatory about TensorDict compatibility

Changed from 'avoid issue #42370' to 'for compatibility with TensorDict and other Mapping subclasses'
so users don't need to look up the issue on GitHub to understand why .keys() is needed.

Addresses maintainer feedback.

pankajbaid567 commented:
The failing test [test_training_gradient_checkpointing_use_reentrant_false](vscode-file://vscode-app/private/var/folders/3r/brgdvkx91hv16l5vj24sd7k00000gn/T/AppTranslocation/11644A3F-AD49-4E85-B517-6AE093C8E246/d/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) in Longformer is unrelated to this PR. Here's why:

Scope of changes: This PR only modifies tokenization padding logic (adding explicit .keys() calls) in 5 tokenizer files. No changes to model architecture or gradient computation.

Luke's gradient checkpointing tests are already skipped: The Luke model (one of the modified tokenizers) has all three gradient checkpointing tests explicitly disabled with @unittest.skip because "This architecture seem to not compute gradients properly when using GC" (see [test_modeling_luke.py](vscode-file://vscode-app/private/var/folders/3r/brgdvkx91hv16l5vj24sd7k00000gn/T/AppTranslocation/11644A3F-AD49-4E85-B517-6AE093C8E246/d/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)).

No connection between changes and failure:

Tokenization happens during data preprocessing
Gradient checkpointing happens during model forward/backward pass
These are completely separate phases with no interaction

Rocketknight1 commented:
Hmn, I'd still remove the tests and `is_tensordict_available`! It's not a dependency anywhere, so our CI is unlikely to ever have it installed. As a result, those tests are just going to get skipped anyway.

baidpankaj567 committed:
Remove TensorDict tests and utilities as requested

- Removed TensorDictCompatibilityTest class from test_data_collator.py
- Removed is_tensordict_available() and require_tensordict() from testing_utils.py
- TensorDict is not a CI dependency, so these tests would be skipped anyway
- The .keys() fix for TensorDict compatibility remains in place

pankajbaid567 commented:
@Rocketknight1 I have made the required changes sir

Rocketknight1 reviewed (approved):
Yep. LGTM now, thank you!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_42373). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
